streamflow/pystreamflow/model.py

```diff
@@ -2,41 +2,285 @@
 """
 
 import math
+import numpy as np
+import scipy.integrate
+import pints
+import pystreamflow
 
+# Try to import the optional scikit odes solver, and record whether or not it
+# was available
+try:
+    import scikits.odes
+    scikit_ode_avail = True
+except ImportError:
+    scikit_ode_avail = False
 
-def flux(S, a):
-    r"""Flux function for hydrological processes.
 
-    This is a general function for the relative flux as it depends on relative
-    storage and a single shape parameter.
+class RiverModel(pints.ForwardModel):
+    """Rainfall runoff model of river streamflow.
 
-    The equation is given by
+    The model has four latent state variables:
 
-    .. math::
-        f(S, a) = \frac{1 - e^{-aS}}{1 - e^{-a}}
+        * S_i = interception storage
+        * S_u = unsaturated storage
+        * S_s = slow reservoir
+        * S_f = fast reservoir
 
-    To prevent overflow, values of -a*S or -a which exceed 600 will be
-    truncated to 600. This limit should not be reached with typical values of a
-    and S, but it helps to protect the function during inference if strange
-    values are provided by some algorithm. Confer
-    https://github.com/Zaijab/DREAM/blob/master/examples/example_14/crr_model.c
+    and one observed variable:
 
-    Parameters
+        * z = water flowed out of the river
+
+    (In fact, it is dz/dt, or the streamflow, that is observed.)
+
+    The model is characterized by the following unknown parameters:
+
+        * I_max = maximum interception
+        * S_u,max = unsaturated storage capacity
+        * Q_s,max = maximum percolation
+        * alpha_e = evaporation flux shape
+        * alpha_f = runoff flux shape
+        * K_s = slow reservoir time constant
+        * K_f = fast reservoir time constant
+
+    as well as the following two parameters whose values are here assumed fixed
+    and known:
+
+        * alpha_s = 0 (percolation flux shape)
+        * alpha_i = 50 (interception flux shape)
+
+    Appearing multiple times in the model is the flux function f, given by:
+
+        * f(S, a) = (1 - exp(-a * S)) / (1 - exp(-a))
+
+    The behavior of all the variables is governed by a system of differential
+    equations, namely
+
+        * dS_i/dt = Precip(t) - InterceptEvap(t) - EffectPrecip(t)
+        * dS_u/dt = EffectPrecip(t) - UnsatEvap(t) - Percolation(t) - Runoff(t)
+        * dS_s/dt = Percolation(t) - SlowStream(t)
+        * dS_f/dt = Runoff(t) - FastStream(t)
+        * dz/dt = SlowStream(t) + FastStream(t)
+
+    Each term is defined below.
+
+    Precip = measured precipitation (provided as input to the model)
+
+    Evap = measured or theoretical evaporation (provided as input to the model)
+
+    InterceptEvap = evaporation from the interception component
```

Comment:
What does interception component mean?

Reply:
I added some extra explanations for these terms (also for the two comments below)